dhall/src/Dhall/TH.hs

```diff
@@ -22,28 +66,209 @@
+                    , "Dhall type:                                                                     \n"
+                    , "                                                                                \n"
+                    , " " <> Dhall.Util.insert dhallType <> "\n"
+                    , "                                                                                \n"
+                    , "... where a simpler type was expected."
+                    ]
+
+            let message = Pretty.renderString (Dhall.Pretty.layout document)
+
+            fail message
+
+-- | Convert a Dhall type to the corresponding Haskell constructor type
+toConstructor :: Pretty a => (Text, Maybe (Expr s a)) -> Q Con
+toConstructor (constructorName, maybeAlternativeType) = do
+    let name = Syntax.mkName (Text.unpack constructorName)
+
+#if MIN_VERSION_template_haskell(2,11,0)
+    let bang = Bang NoSourceUnpackedness NoSourceStrictness
+#else
+    let bang = NotStrict
+#endif
+
+    case maybeAlternativeType of
+        Just (Record kts) -> do
+            let process (key, dhallFieldType) = do
+                    haskellFieldType <- toSimpleHaskellType dhallFieldType
+
+                    return (Syntax.mkName (Text.unpack key), bang, haskellFieldType)
+
+            varBangTypes <- traverse process (Dhall.Map.toList kts)
+
+            return (RecC name varBangTypes)
+
+        Just dhallAlternativeType -> do
+            haskellAlternativeType <- toSimpleHaskellType dhallAlternativeType
+
+            return (NormalC name [ (bang, haskellAlternativeType) ])
+
+        Nothing -> do
+            return (NormalC name [])
+
+-- | Generate a Haskell datatype declaration from a Dhall type
+--
+-- This comes in handy if you need to keep a Dhall type and Haskell type in
+-- sync.  You make the Dhall type the source of truth and use Template Haskell
+-- to generate the matching Haskell type declaration from the Dhall type.
+--
+-- For example, this Template Haskell splice:
+--
+-- > Dhall.TH.makeHaskellType "T" "< A : { x : Bool } | B >"
+--
+-- ... generates this Haskell code:
+--
+-- > data T = A {x :: GHC.Types.Bool} | B
+--
+-- To add any desired instances (such as `Dhall.FromDhall`/`Dhall.ToDhall`),
+-- you can use the `StandaloneDeriving` language extension, like this:
+--
+-- > {-# LANGUAGE DeriveAnyClass     #-}
+-- > {-# LANGUAGE DeriveGeneric      #-}
+-- > {-# LANGUAGE OverloadedStrings  #-}
+-- > {-# LANGUAGE StandaloneDeriving #-}
+-- > {-# LANGUAGE TemplateHaskell    #-}
+-- >
+-- > Dhall.TH.makeHaskellType "T" "< A : { x : Bool } | B >"
+-- > 
+-- > deriving instance Generic   T
+-- > deriving instance FromDhall T
+--
+-- Carefully note that the generated Haskell type only corresponds to the Dhall
+-- type when using the `Dhall.Smart` setting for the
+-- `Dhall.singletonConstructors` field of `InterpretOptions`.  That setting
+-- will eventually become the default, but until then you will need to
```

Comment:
How about just changing the default right now, to simplify these docs?!

Reply:
Good idea!  I'll go ahead and change the default